WalletWasabi/Services/Terminate/TerminateService.cs

```diff
@@ -0,0 +1,90 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using WalletWasabi.Logging;
+
+namespace WalletWasabi.Services.Terminate
+{
+	public class TerminateService
+	{
+		private Func<Task> _terminateApplicationAsync;
+
+		private const long TerminateStatusIdle = 0;
+		private const long TerminateStatusInProgress = 1;
+		private const long TerminateFinished = 2;
+
+		private long _terminateStatus;
+
+		public TerminateService(Func<Task> terminateApplicationAsync)
+		{
+			_terminateApplicationAsync = terminateApplicationAsync;
+			AppDomain.CurrentDomain.ProcessExit += CurrentDomain_ProcessExit;
+			Console.CancelKeyPress += Console_CancelKeyPress;
+		}
+
+		private void CurrentDomain_ProcessExit(object? sender, EventArgs e)
+		{
+			Logger.LogDebug("ProcessExit was called.");
+
+			// This must be a blocking call because after this the OS will terminate Wasabi process if exists.
+			Terminate();
+		}
+
+		private void Console_CancelKeyPress(object? sender, ConsoleCancelEventArgs e)
+		{
+			Logger.LogWarning("Process was signaled for termination.");
```

Comment:
```suggestion
			Logger.LogWarning($"Process termination was requested using '{e.SpecialKey}' keyboard shortcut.");
```

Possibly one can log also the reason why.